### PR TITLE
Reject request if NUL is present in the request line

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We should not allow extra whitespaces or control chars in the request line.

Modifications:

- Check if the there are extra chars at the end.
- Add unit test

Result:

Fix https://github.com/netty/netty/issues/14166

### CODE REVIEW

**Review by franz1981 — APPROVED** [review]

**Comment by vietj** [comment]
LGTM

**Review by violetagg — APPROVED** [review]

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java [hunk 1]
```diff
@@ -948,6 +948,12 @@ private String[] splitInitialLine(ByteBuf asciiBuffer) {
 
         final int end = startContent + asciiBuffer.readableBytes();
 
+        if (isControlOrWhitespaceAsciiChar(asciiBytes[end - 1])) {
+            // There should no extra control or whitespace char.
+            // See https://datatracker.ietf.org/doc/html/rfc2616#section-5.1
+            throw new IllegalArgumentException();
```
Comment (bryce-anderson) [inline codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java:954]: ffti:
```suggestion
            throw new IllegalArgumentException("Illegal character in request line: 0x" + Integer.ToHexString(char));
```

**Review by bryce-anderson — APPROVED** [review]

File: codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java (hunk 1 above)
Comment (idelpivnitskiy) [reply codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java:954]: +1 for a hex representation and index of its position

Comment (normanmaurer) [reply codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectDecoder.java]: Good idea!

**Comment by violetagg** [comment]
@normanmaurer I did some extended testing and it seems that this fix violates the specification. For example Tomcat returns the following:

```
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 48 54 54 50 2f 31 2e 31 20 32 30 30 20 0d 0a 54 |HTTP/1.1 200 ..T|
|00000010| 72 61 6e 73 66 65 72 2d 45 6e 63 6f 64 69 6e 67 |ransfer-Encoding|
|00000020| 3a 20 63 68 75 6e 6b 65 64 0d 0a 44 61 74 65 3a |: chunked..Date:|
|00000030| 20 54 75 65 2c 20 31 36 20 4a 75 6c 20 32 30 32 | Tue, 16 Jul 202|
|00000040| 34 20 30 37 3a 32 33 3a 33 32 20 47 4d 54 0d 0a |4 07:23:32 GMT..|
|00000050| 0d 0a 32 33 0d 0a 68 74 74 70 3a 2f 2f 5b 3a 3a |..23..http://[::|
|00000060| 31 5d 3a 36 32 36 32 39 2f 3f 71 3d 74 65 73 74 |1]:62629/?q=test|
|00000070| 25 32 30 64 25 32 30 64 71 0d 0a 30 0d 0a 0d 0a |%20d%20dq..0....|
+--------+-------------------------------------------------+----------------+
```

With the latest changes this response leads to

```
   java.lang.IllegalArgumentException: Illegal character in request line: 0x20
        at io.netty.handler.codec.http.HttpObjectDecoder.splitInitialLine(HttpObjectDecoder.java:955)
        at io.netty.handler.codec.http.HttpObjectDecoder.decode(HttpObjectDecoder.java:351)
```

According to https://datatracker.ietf.org/doc/html/rfc9112#name-status-line the space is expected and not optional

```
status-line = HTTP-version SP status-code SP [ reason-phrase ]
```

**Comment by normanmaurer** [comment]
@violetagg let me fix it

**Pull request merged**